Enhances hierarchy violation report within PhaseCheckHierarchy phase

pluveto opened the pull request:
# Context, Motivation & Description

This change enhances the error reporting capabilities within the `PhaseCheckHierarchy` phase.

Previously, when a hierarchy violation (e.g., illegal cross-component signal assignment or read) or a related issue like "OLD NETLIST RE-USED" occurred, the error messages provided by `PhaseCheckHierarchy` were often generic. This made it challenging for developers especially beginners to quickly identify the precise nature of the error. 

This modification refactors the hierarchy checking logic to provide significantly more detailed and actionable diagnostic messages.

# Impact on code generation

Zero. Solely improves compile-time diagnostic messages.

# Checklist

- [N/A] Unit tests were added (no need).
- [N/A] API changes are or will be documented (no public API alteration).

pluveto commented:
The new error report now looks like:

```
[info] HIERARCHY VIOLATION (Assignment): Signal '(toplevel/freeList/io_currentState_freeMask : out Bits[4 bits])' (an instance of Bits), defined in component 'freeList' (parent: 'toplevel') with direction 'output',
[info]   is assigned by expression '(B"1111" 4 bits)', but is not assignable from component 'toplevel' (parent: 'None').
[info] To be assignable from component 'toplevel' (parent: 'None'), signal '(toplevel/freeList/io_currentState_freeMask : out Bits[4 bits])' must satisfy one of the following:
[info]   1. Be a directionless signal defined directly within component 'toplevel' (parent: 'None') (Actual: false).
[info]   2. Be an 'out' or 'inout' port of component 'toplevel' (parent: 'None') (Actual: false).
[info]   3. Be an 'in' or 'inout' port of a direct child component of component 'toplevel' (parent: 'None') (Actual: false).
[info] Contextual Details:
[info]   - Target Signal ('(toplevel/freeList/io_currentState_freeMask : out Bits[4 bits])'):
[info]     - Full Hierarchical Name: freeList
[info]     - Defining Component: component 'freeList' (parent: 'toplevel')
[info]     - isDirectionLess: false, isInput: false, isOutput: true, isInOut: false
[info]   - Current Component (where assignment occurs):
[info]     - component 'toplevel' (parent: 'None')
[info] Location of assignment:     at parallax.test.rename.RenameUnitTestBench.<init>(RenameUnitSpecV2.scala:51)
[info]     at parallax.test.rename.RenameUnitSpec.$anonfun$new$32(RenameUnitSpecV2.scala:749)
```

mrcmry commented:
Very nice improvement !

Do you think it's necessary to repeat the signal full signal name (which can be very long) ? What about something like:

```
To be assignable from component 'toplevel' (parent: 'None'), this signal must satisfy one of the following:
```

```
Contextual details for this target signal:
     - Defining Component: component 'freeList' (parent: 'toplevel')
```

Also if I understand correctly the "(Actual: false)" is kind of redundant because if one of the condition would be met, the error would not occur. So perhaps we could write something like:

```
To be assignable from component 'toplevel' (parent: 'None'), this signal must satisfy one of the following:
  1. Be a directionless signal defined directly within component 'toplevel'
  2. Be an 'out' or 'inout' port of component 'toplevel' 
  3. Be an 'in' or 'inout' port of a direct child component of component 'toplevel'
```

Regarding the scala call stack, it's important to have the "    at" at the beginning of the line so IDE like vscode detect then like scala/java call stack like this:

```
[info] Location of assignment:     
[info]     at parallax.test.rename.RenameUnitTestBench.<init>(RenameUnitSpecV2.scala:51)
[info]     at parallax.test.rename.RenameUnitSpec.$anonfun$new$32(RenameUnitSpecV2.scala:749)
```
See this PR #1698 for an example with screenshoots.

finally I'm probably not the best person to point this out (my native language is French, I do speak German as well), but I think capitalization of the first letter of the word is reserved for titles so I think "Target Signal" -> "Target signal", etc.

I just did a bit of research about capitalization in English and found this:
https://writer.com/blog/capitalization-rules/
https://en.wikipedia.org/wiki/Title_case
https://www.purchase.edu/editorial-style-guide/general-style-preferences/punctuation/bulleted-and-numbered-lists/

If we apply these rules your original message would like this:
```
[info] HIERARCHY VIOLATION (assignment): signal '(toplevel/freeList/io_currentState_freeMask : out Bits[4 bits])' (an instance of Bits), defined in component 'freeList' (parent: 'toplevel') with direction 'output',
[info]   is assigned by expression '(B"1111" 4 bits)', but is not assignable from component 'toplevel' (parent: 'None').
[info] To be assignable from component 'toplevel' (parent: 'None'), signal '(toplevel/freeList/io_currentState_freeMask : out Bits[4 bits])' must satisfy one of the following:
[info]   1. Be a directionless signal defined directly within component 'toplevel' (parent: 'None') (actual: false).
[info]   2. Be an 'out' or 'inout' port of component 'toplevel' (parent: 'None') (actual: false).
[info]   3. Be an 'in' or 'inout' port of a direct child component of component 'toplevel' (parent: 'None') (actual: false).
[info] Contextual details:
[info]   - Target signal ('(toplevel/freeList/io_currentState_freeMask : out Bits[4 bits])'):
[info]     - Defining component: component 'freeList' (parent: 'toplevel')
[info]     - isDirectionLess: false, isInput: false, isOutput: true, isInOut: false
[info]   - Current component (where assignment occurs):
[info]     - component 'toplevel' (parent: 'None')
[info] Location of assignment:     at parallax.test.rename.RenameUnitTestBench.<init>(RenameUnitSpecV2.scala:51)
[info]     at parallax.test.rename.RenameUnitSpec.$anonfun$new$32(RenameUnitSpecV2.scala:749)
```

But I think it's not very important. I suppose your native language do not have capital letter, even for me English capitalization is very complicated. If you prefer don't spend to much time with capitalization, I can do further PR later to fix this.

Overall this improvement of the error message is very important in my opinion. I'm glad you started this improvement process.

pluveto commented:
OK I'd like to refine this PR. Working on it.

pluveto commented:
Thank you for the very detailed feedback! I've incorporated your suggestions and updated the PR. Now the error report looks like:

```
[info] HIERARCHY VIOLATION (assignment):
[info]   Attempted to assign to signal '(toplevel/io_uopIn_valid : in Bool)' (an instance of Bool),
[info]   which is defined in component 'toplevel' with direction 'input',
[info]   but is not assignable from component 'toplevel'.
[info]   (Assigned by expression: '(toplevel/alu/io_uopIn_valid : in Bool)')
[info] To be assignable from component 'toplevel', this signal must satisfy one of the following:
[info]   1. Be a directionless signal defined directly within component 'toplevel'.
[info]   2. Be an 'out' or 'inout' port of component 'toplevel'.
[info]   3. Be an 'in' or 'inout' port of a direct child component of component 'toplevel'.
[info] Location of assignment:
[info]     at parallax.test.scala.execute.DemoAluTestBench.<init>(DemoAluSpec.scala:50)
[info]     at parallax.test.scala.execute.DemoAluSpec.$anonfun$new$2(DemoAluSpec.scala:386)
```
